runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -12268,50 +12346,72 @@ void generateFillInDataBlockSequenceForUnresolvedField(TR::Node *node, J9Method*
-
-   TR::RegisterDependencyConditions  *deps = generateRegisterDependencyConditions(1, 1, cg);
    deps->addPreCondition(dataBlockReg, TR::RealRegister::NoReg, cg);
    deps->addPostCondition(dataBlockReg, TR::RealRegister::NoReg, cg);
-   deps->stopAddingConditions();
-   generateLabelInstruction(LABEL, node, endLabel, deps, cg);
+
+   generateLabelInstruction(LABEL, node, startLabel, cg);
+   generateRegMemInstruction(LEARegMem(), node, dataBlockReg, generateX86MemoryReference(dataSnippet, cg), cg);
+   generateMemImmInstruction(CMPMemImms(), node, generateX86MemoryReference(dataBlockReg, offsetInDataBlock, cg), -1, cg);
+   generateLabelInstruction(JE4, node, unresolveLabel, cg);
 
       {
-      TR_OutlinedInstructionsGenerator og(unresolvedLabel, node ,cg);
+      TR_OutlinedInstructionsGenerator og(unresolveLabel, node ,cg);
       if (isStatic)
          {
          // Fills in J9JITWatchedStaticFieldData.fieldClass
-         TR::Register *fieldClassReg = cg->evaluate(isWrite ? node->getSecondChild()->getFirstChild(): node->getFirstChild());
+         TR::Register *fieldClassReg = cg->evaluate(isWrite ? getDirectWrtbarJ9ClassNode(node): node->getFirstChild());
          generateMemRegInstruction(SMemReg(is64Bit), node, generateX86MemoryReference(dataBlockReg, (intptrj_t)(offsetof(J9JITWatchedStaticFieldData, fieldClass)), cg), fieldClassReg, cg);
+         deps->addPreCondition(fieldClassReg, TR::RealRegister::NoReg, cg);
+         deps->addPostCondition(fieldClassReg, TR::RealRegister::NoReg, cg);
          }
 
-      //call resolving helper
       TR::ResolvedMethodSymbol *methodSymbol = node->getByteCodeInfo().getCallerIndex() == -1 ? cg->comp()->getMethodSymbol(): cg->comp()->getInlinedResolvedMethodSymbol(node->getByteCodeInfo().getCallerIndex());
-      TR::Node *cpAddressNode = TR::Node::createWithSymRef(node, TR::loadaddr, 0, cg->comp()->getSymRefTab()->findOrCreateConstantPoolAddressSymbolRef(methodSymbol));
-      TR::Node *cpIndexNode = TR::Node::iconst(node, symRef->getCPIndex());
-      TR::Node *helperCallNode = TR::Node::createWithSymRef(TR::acall, 2, 2, cpIndexNode, cpAddressNode,
-            cg->comp()->getSymRefTab()->findOrCreateRuntimeHelper(helperIndex, true /* canGCandReturn */, true /*canGCandExcept*/, false /* preservesAllRegisters*/));
-      helperCallNode->incReferenceCount();
-      TR::Register *resultReg = TR::TreeEvaluator::performCall(helperCallNode, false /* isIndirect */, false /* spillFPRegs */, cg);
+      if (is64Bit)
+         {
+         TR::Register *cpAddressReg = cg->allocateRegister();
+         TR::Register *cpIndexReg = cg->allocateRegister();
+         generateRegImm64SymInstruction(MOV8RegImm64, node, cpAddressReg, (uintptrj_t) methodSymbol->getResolvedMethod()->constantPool(), cg->comp()->getSymRefTab()->findOrCreateConstantPoolAddressSymbolRef(methodSymbol), cg);
+         generateRegImm64Instruction(MOV8RegImm64, node, cpIndexReg, (uintptrj_t)symRef->getCPIndex(), cg);
```

Comment:
`getCPIndex()` has value type `int32_t`, it seems we can safely `MOV8RegImm4` unless it need be zero-extended to a 64-bit value instead of sign-extension.

Reply:
Thanks for the suggestion. Changed accordingly.